pkg/frontend/transport/handler.go

```diff
@@ -252,6 +264,80 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getQueryDetail(url string, values url.Values) QueryDetail {
+	urlValues := parseURLValues(values)
+	return QueryDetail{
+		QueryString: urlValues.query,
+		StepInt:     urlValues.stepInt,
+		StartInt:    urlValues.startInt,
+		EndInt:      urlValues.endInt,
+		TsInt:       urlValues.tsInt,
+	}
+}
+
+type urlValues struct {
+	query                            string
+	stepInt, startInt, endInt, tsInt int64
+}
+
+func parseURLValues(values url.Values) urlValues {
+	query := values.Get("query")
+	step := values.Get("step")
+	start := values.Get("start")
+	end := values.Get("end")
+	ts := values.Get("time")
+	var (
+		tsInt, startInt, endInt, stepInt int64
+		err                              error
+	)
+	if len(step) == 0 { // instance query
+		if len(ts) > 0 {
+			tsInt, err = util.ParseTime(ts)
+			if err != nil {
+				util_log.Logger.Log("msg", "failed to parse time", "err", err)
+			}
+		} else {
+			tsInt = util.TimeToMillis(time.Now())
+		}
+		return urlValues{
+			query: query,
+			tsInt: tsInt,
+		}
+	}
+
+	//range query
+	stepInt, err = util.ParseDurationMs(step)
+	if err != nil {
+		util_log.Logger.Log("msg", "failed to parse step", "err", err)
+	}
+	if len(start) > 0 {
+		startInt, err = util.ParseTime(start)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse start", "err", err)
+		}
+	}
+	if len(end) > 0 {
+		endInt, err = util.ParseTime(end)
+		if err != nil {
+			util_log.Logger.Log("msg", "failed to parse end", "err", err)
+		}
+	}
+	return urlValues{
+		query:    query,
+		startInt: startInt,
+		endInt:   endInt,
+		stepInt:  stepInt,
+	}
+}
+
+type QueryDetail struct {
```

Comment:
I think we are not using this struct at all? And to me it is the same as `urlValues`. That one seems also redundant.
Can you just pass `stats` into `parseURLValues` and set values there?